crates/transaction-pool/src/maintain.rs

```diff
@@ -0,0 +1,268 @@
+                //
+                // Note: we no longer know if the tx was local or external
+                let _ = pool.add_external_transactions(pruned_old_transactions).await;
+                // TODO: metrics
+            }
+            CanonStateNotification::Commit { new } => {
+                // TODO skip large commits?
+
+                let (blocks, state) = new.inner();
+                let tip = blocks.tip();
+                // base fee for the next block: `tip+1`
+                let pending_block_base_fee = calculate_next_block_base_fee(
+                    tip.gas_used,
+                    tip.gas_limit,
+                    tip.base_fee_per_gas.unwrap_or_default(),
+                ) as u128;
+
+                let first_block = blocks.first();
+                // check if the range of the commit is canonical
+                if first_block.parent_hash == pool_info.last_seen_block_hash {
+                    let changed_accounts = changed_accounts_iter(state).collect();
+                    let mined_transactions = blocks.transactions().map(|tx| tx.hash).collect();
+                    // Canonical update
+                    let update = CanonicalStateUpdate {
+                        hash: tip.hash,
+                        number: tip.number,
+                        pending_block_base_fee,
+                        changed_accounts,
+                        mined_transactions,
+                    };
+                    pool.on_canonical_state_change(update);
+                } else {
+                    // TODO is this even reachable, because all commits are canonical?
+                    // this a canonical
+                }
+            }
+        }
+    }
+}
+
+/// A unique ChangedAccount identified by its address that can be used for deduplication
+#[derive(Eq)]
+struct ChangedAccountEntry(ChangedAccount);
+
+impl PartialEq for ChangedAccountEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.address == other.0.address
+    }
+}
+
+impl Hash for ChangedAccountEntry {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.address.hash(state);
+    }
+}
+
+impl Borrow<Address> for ChangedAccountEntry {
+    fn borrow(&self) -> &Address {
+        &self.0.address
+    }
+}
+
+/// Loads all accounts at the given state
+///
+/// Returns an error with all given addresses if the state is not available.
+fn load_accounts<Client, I>(
+    client: &Client,
+    at: BlockHash,
+    addresses: I,
+) -> Result<Vec<ChangedAccount>, Box<(HashSet<Address>, reth_interfaces::Error)>>
+where
+    I: Iterator<Item = Address>,
+
+    Client: StateProviderFactory,
+{
+    let mut changed_accounts = Vec::new();
+    let state = match client.history_by_block_hash(at) {
+        Ok(state) => state,
+        Err(err) => return Err(Box::new((addresses.collect(), err))),
+    };
+    for addr in addresses {
+        if let Ok(maybe_acc) = state.basic_account(addr) {
```

Comment:
why is the error ignored here? seems like the addresses that errored wouldn't be added to dirty

Reply:
good point